lib/signatures.ts

```diff
@@ -1,16 +1,24 @@
 import { promises } from 'fs';
 import { SignatureResult } from './types';
 import { getDubHashSignature } from './dubhash';
+import { getUhashSignature } from './uhash';
 
 const { readFile } = promises;
 
 export async function getSignature(
   filePath: string,
+  hashType = 'dubhash',
 ): Promise<SignatureResult | null> {
   try {
     const fileContents = await readFile(filePath);
     if (fileContents.length === 0) return null;
-    return getDubHashSignature(filePath, fileContents);
+    if (hashType === 'dubhash') {
+      return getDubHashSignature(filePath, fileContents);
+    }
+    if (hashType === 'uhash') {
```

Comment:
else if maybe? or maybe a switchcase?

Reply:
Agreed - a switch could be good if there's potential to add more hash types in future?

Reply:
Side note - was just looking at the signature specification docs, and it seems to jump from format 1 (double hash) to format 3 (uhash). Is there a different format in the middle that I've missed or something, or was it just a typo? https://github.com/snyk/snyk-cpp-plugin/blob/master/docs/signature_specification.md#full-file-hashes

Reply:
Format 2 is a SHA1 of the full file needed only to be compliant with the SPDX spec when generating SPDX-reports. And since we're not doing this yet in the c/c++ use-case we left it out from the specification, resulting in this weird jump from 1 to 3.

Reply:
This file could be updated to

```
import { promises } from 'fs';
import { SignatureHashAlgorithm, SignatureResult } from './types';
import { getDubHashSignature } from './dubhash';
import { getUhashSignature } from './uhash';
import pMap = require('p-map');

const { readFile } = promises;

export async function getSignaturesByAlgorithm(
  filePaths: string[],
  hashType: SignatureHashAlgorithm = 'dubhash',
): Promise<(SignatureResult | null)[]> {
  if (hashType !== 'dubhash' && hashType !== 'uhash') {
    throw new Error(`Unsupported hashType ${hashType}`);
  }
  let signatureMapperToBeUsed = getDubHashSignatureMapper;
  if (hashType === 'uhash') {
    signatureMapperToBeUsed = getUHashSignatureMapper;
  }
  return await pMap(filePaths, signatureMapperToBeUsed, { concurrency: 20 });
}

async function getDubHashSignatureMapper(
  filePath: string,
): Promise<SignatureResult | null> {
  const fileContents = await readFile(filePath);
  if (fileContents.length === 0) return null;
  return getDubHashSignature(filePath, fileContents);
}

async function getUHashSignatureMapper(
  filePath: string,
): Promise<SignatureResult | null> {
  const fileContents = await readFile(filePath);
  if (fileContents.length === 0) return null;
  return getUhashSignature(filePath, fileContents);
}
```

This way allow us of upfront decide which algorithm we want to use and easily adopt the require `mapper` to be passed down to `p-map` function to concurrently read and generate signatures from all the existing filePaths

Reply:
Did you test this one? I couldn't get it to work. I believe mapper needs to have the call signature `mapper(element, index)`, so adding a second argument `hashType` is not possible (unless it were somehow wrapped into element). So if we want to use dubhash or uhash without any code changes, we need to expose something like `getSignatureMapper(hashType)` that would return a mapper function based on the specified hashType.

Reply:
I would recommend to take a look at the other points, puzzling all of them together makes it work.
The idea of have a function that deals with return the kind of mapper that should be used rather 
than an if as I've suggested above, sounds good as alternative as the number of mappers/algorithms may increase overtime